### ISSUE TITLE
Graphic tasks

### ISSUE THREAD

**Issue by piranna** [NodeOS/NodeOS#201]
Since there's so much interest on having a high level GUI running locally on NodeOS I will list here the tasks that need to be done to make this feasable:

[ ] Add EGL support to NodeOS
  [ ] Compile libMesa without any dependency on X11
[ ] Webkit running on framebuffer (DRI)
[ ] WebKit running on EGL
[ ] Static build of Webkit Context_Renderer
  [ ] Add support to use full screen resolution, not default 800x600
  [ ] Update resolution on framebuffer resize (Is that possible?)
[ ] Add WebGL support to node-canvas (OpenGL ES 2.0)

Anything more?

**Comment by lite20** [comment]
Yes the frame buffer size can be updated. I've done so much research I can quote code from the kernel XD

This looks like an about complete list to get started. Would you like me to make a fork for us to work on instead of throwing things at the main thread? I do remember you suggesting that the graphic edition be separated from the server edition?

**Comment by piranna** [comment]
> I've done so much research I can quote code from the kernel XD

Ah, that's good :-P


> Would you like me to make a fork for us to work on instead of throwing things at the main thread?

I think this things can be worked and tested on regular Linux systems, so I would work on them as independent projects and later see how to integrate them (maybe also published as npm packages... :-P)

**Comment by Coretool** [comment]
Could you add "/gecko" where you put webkit ? ^^

**Comment by piranna** [comment]
> Could you add "/gecko" where you put webkit ? ^^

Done ;-)

**Comment by lite20** [comment]
> I think this things can be worked and tested on regular Linux systems, so I would work on them as independent projects and later see how to integrate them (maybe also published as npm packages... :-P)

Yeah this was my intention. Fourth npm module, here I come!

(Does anyone else realize how our efforts for this project are pushing other things along in other projects at the same time?)

**Comment by piranna** [comment]
> Does anyone else realize how our efforts for this project are pushing other things along in other projects at the same time?

That's just one of the thing I mostly love of Node.js and npm and why I'm trying to make NodeOS elements as standard and reusable on other systems as possible :-D

**Comment by lite20** [comment]
> That's just one of the thing I mostly love of Node.js and npm and why I'm trying to make NodeOS elements as standard and reusable on other systems as possible :-D

NodeOS = cure to lyfe

**Comment by Compynerd255** [comment]
A thought about graphics integration: Something that I am reminded of for this project is my old Apple //e (as I mentioned in another thread), which had graphics [but also supported 4 lines of text at the bottom of the screen](http://www.virtualapple.org/img/AbyssalZone.jpg), which could be turned on or off. We could perhaps take similar inspiration at least in the development stages of NodeOS, where most of the screen is a graphical environment but several lines of text are found at the bottom where the command-line user still interacts as they did before. The advantage to this would be that we don't have to implement both graphics and hitbox input at the same time.

**Comment by lite20** [comment]
@compynerd255
> most of the screen is a graphical environment but several lines of text are found at the bottom where the command-line user still interacts as they did before.

I actually really like your suggestion. NodeOS doesn't have a default window manager as graphics aren't a part of the core operating system.

Because of this, each window manager would have to implement it on their own. now, I can forsee a lot of managers being built of a rendering engine u've been developing (and now testing :D) called Yunitto, and I will attempt to bake this feature into it as a configurable addition making it easier for wm developers to integrate.

Awesome suggestion!

**Comment by Compynerd255** [comment]
@lite20 I'm glad that you like it. Of course, I'm not sure how useful this feature would be for general public use - it might be easier on us if we just wrote in a mouse. I can certainly see this feature being useful for development, though - the text, which could also be on the top or the side of the screen, could be used as a debugging log / interaction window. It could even be used as a simple HUD. I'll post these ideas in more detail on your Yunitto repository.

**Comment by lite20** [comment]
>  I'll post these ideas in more detail on your Yunitto repository.

Yes thank you! This way I can ensure it's executed and properly developed.

> Of course, I'm not sure how useful this feature would be for general public use - it might be easier on us if we just wrote in a mouse.

For general purpose users, yes this is probably  a weird window they may not be a fan of, but it could be incredibly useful to developers, and if we made the source swappable, any app could tap in and use it.

**Comment by piranna** [comment]
Isn't this similar to the browser dev console? Also, this could be easily done with blessed or a div, I don't think it's needed to have it at the graphic layer but the application one...

**Comment by lite20** [comment]
> Isn't this similar to the browser dev console? 

True. This however isn't accessible on mobile devices unfortunately.

> Isn't this similar to the browser dev console? Also, this could be easily done with blessed or a div, I don't think it's needed to have it at the graphic layer but the application one...

Hm. Good point. What I was thinking was just making a module and apps could use it as they pleased.

**Comment by Compynerd255** [comment]
> Isn't this similar to the browser dev console?

Similar, but not the same - a pure text window offers the following features that a dev console does not:

- The text window can display whatever it wants, and can even move the cursor around and overwrite itself.
- The text window can be interacted with by normal TTY interaction.
- Any type of *input* (not just context sensitive input) can be requested from the text window. (This is probably the most important feature).

One (game) application I could see of this would be an audio-visual version of Zork or another text adventure game: it is played using the text prompt, exactly as the original, but most of the screen is occupied with a visual retelling.

This could also be very useful in other applications if the background could be transparent.

**Comment by lite20** [comment]
> One (game) application I could see of this would be an audio-visual version of Zork or another text adventure game: it is played using the text prompt, exactly as the original, but most of the screen is occupied with a visual retelling.

> This could also be very useful in other applications if the background could be transparent.

With all these customizations this might be a reasonably sized module. I like it!

**Comment by lite20** [comment]
Just a side note too graphic developers, don't forget about this burried thread:

https://github.com/NodeOS/NodeOS/issues/147

**Comment by piranna** [comment]
> Just a side note too graphic developers, don't forget about this burried thread:
> 
> #147

Thanks for remind it, definitely GitHub needs some mechanism to pin some issues on the top, since we are using it somewhat as a forum. Please make them all of you [a suggestion](https://github.com/contact) about that to give some pressure :-) By the moment, I've created [a new label](https://github.com/NodeOS/NodeOS/labels/pinned) so we can "pin" some important issues there :-)

**Comment by lite20** [comment]
Say we want the location of wallpapers to be standard across window managers. Does NodeOS have a location where files and data that are to be shared should be?

We could use voidpool to keep a map of all the files to their respective locations so that shared things like the path to the desktop wallpaper could be fetched with `pool.get('wallpaper')`

**Comment by piranna** [comment]
> Say we want the location of wallpapers to be standard across window managers. Does NodeOS have a location where files and data that are to be shared should be?

I didn't consider this use case... Since each windows manager is considered a standalone project it would have its own wallpapers as resources of itself. To have a common location, I only can think about store them on an independent module that all window managers would have as dependency, or have some mechanism like [XDG user directories](https://wiki.archlinux.org/index.php/Xdg_user_directories) to be able to define some common folders.

Take also in account that the windows manager will run client side and it's independent of the NodeOS backend server, so also it could be a Chrome apps connecting to a remote NodeOS server and have embedded the wallpapers. If you want to have some wallpapers in common, I only can think about a `$HOME/Images/wallpapers` folder or something similar, but user would need to put them there manually. I think windows manager could only be able to show by default the wallpapers that has embedded inside their own module or any of its dependencies, but it's not possible to install them outside itself without a postinstall script, and this could be dangerous because it would overwrite previous ones.

**Comment by lite20** [comment]
Ah okay. The reason I mentioned was because the user might have an app to change the wallpaper. If the wallpaper was in a different location for each window manager, the app would have to have several implementations.

> I only can think about a $HOME/Images/wallpapers folder or something similar

The reason I suggested using Voidspace is so the path could be changed. Yes the user would have to put it there manually but apps could check and if it existed, use it. Else a default of course.

By $HOME are you referring to the users home directory? Just making sure

**Comment by lite20** [comment]
I have an idea!

If anyone has ever created an Electron.js app (I have) they're in the same situation as the apps we build. Electron apps have a "server" layer which operates on the NodeJs thread, and a "browser" layer, which is CEF.

NodeOS all's are basically in the same situation but with a gap potentially between the browser and  server.

To work with this in electron, they introduced a module called "remote". Remote allowed the browser thread to " require" modules on the server side. The server side would do all the execution and return results to the browser thread.

I'm going to look a bit into their code to see how they did it but maybe we could use a similar "remote" module that allowed users to perform operations in the server side and get the results. (RPC, but with more hype :) )

**Comment by piranna** [comment]
> The reason I mentioned was because the user might have an app to change the wallpaper. If the wallpaper was in a different location for each window manager, the app would have to have several implementations.

Shouldn't this be a config option of the Desktop Manager?


> The reason I suggested using Voidspace is so the path could be changed. Yes the user would have to put it there manually but apps could check and if it existed, use it. Else a default of course.

Do you have a link to Voidspace? I can't be able to find it anywhere...


> By $HOME are you referring to the users home directory? Just making sure

Yes, that's it.

**Comment by lite20** [comment]
> Shouldn't this be a config option of the Desktop Manager?

Maybe, we've never discussed this before, so it hasn't been standardized. If it would be, we ought to decide how apps would access it. 

If all apps stored their public variables in one location or via one system, fetching something like the desktop wallpaper path. Voidspace could probably do this too.

> Do you have a link to Voidspace? I can't be able to find it anywhere...

It's out on NPM and its repository is on bitbucket.
Npm: https://www.npmjs.com/package/voidspace
Repo: https://bitbucket.org/Lite20/voidspace

The status says beta but the latest version is actually a stable release. I just haven't updated the text :P

Voidspace works on all platforms, including the web.

**Comment by derhuerst** [comment]
Why not make the NodeOS server more independent from the GUI client? I'd like to use NodeOS without any GUI. Also, i'd imagine window managers / desktop environments without anything like wallpapers at all. So please don't narrow the architecture down on the usual mainstream-like GUIs.

That's also the reason why I think that the NodeOS server should know nothing about the GUI client.

**Comment by piranna** [comment]
@derhuerst, I absolutely agree with you, this things should be only client-side. There would be some support modules used by the backend server of the windows managers to have some common features, but definitely this functionality should not be on the NodeOS core. A wallpaper is usually a simple image, and that could be stored on the user browser `localStorage` as a `data:` URI.

**Comment by derhuerst** [comment]
Why not let the window manager / GUI serve itself?

> A wallpaper is usually a simple image, and that could be stored on the user browser localStorage as a data: URI.

AFAIK, browser extensions can have their own bundled files. For other things, a local HTTP server could be started, serving all the assets the window manager needs.

**Comment by lite20** [comment]
> Why not make the NodeOS server more independent from the GUI client?

From what I know, it is. All installs of NodeOS (except maybe pre-installs ;) ) use NodeOS server, and a GUI can be installed as any other npm package.

You install, and bam; GUI. :

> Why not let the window manager / GUI serve itself?

I agree. The interaction between apps and the server doesn't need to be standardized. It's only the interface apps would use that would need to be standard in order to avoid having to recode apps.

**Comment by Coretool** [comment]
Would it be ok to use qt-embedded to use webkit as html rendering engine ? 
Qt-embedded is indeed lighweight and we could use its webkit port... However, I am not sure if this would be against the 'no-gui-framework'-idea (thats why I' m asking)

**Comment by piranna** [comment]
> Would it be ok to use qt-embedded to use webkit as html rendering engine
?
> Qt-embedded is indeed lighweight and we could use its webkit port...
However, I am not sure if this would be against the 'no-gui-framework'-idea
(thats why I' m asking)

A little bit, but as far as it can be compiled statically and work
fullscreen directly on the framebuffer or EGL, it could be a valid first
aproximation.

**Comment by Coretool** [comment]
Another possibility is the CEF. It has got the ability to render to the frambuffer by default which makes it also a legitim possibility

Last but not least we got the good old gecko... I found a b2g port for 'normal' linux versions, however it may be outdated (which is sad but true :/)

1,2 or 3 ? (My personal favourite is 2..)

**Comment by lite20** [comment]
> Another possibility is the CEF. It has got the ability to render to the frambuffer by default which makes it also a legitim possibility

I like #2 :3

**Comment by piranna** [comment]
It has been discussed before, CEF/content_shell over EGL is the desired renderer, but until we get that other alternatives are welcome, like content_shell over framebuffer or Gecko.

**Comment by formula1** [comment]
https://github.com/formula1/node-webkit-canvas

So this is something mentioned before but I've finally gotten around to it.
HTMLDoc -> Processor -> RGB(A?) Values

Using Webkit.js and headless-gl (it no longer uses 2d-contexts from what I can tell)

Ok, so heres where I'm at with this. I am getting no errors within webkit.js however I am getting buffers with all white pixels (255, 255, 255, 255). This may be due to the viewport not being handled correctly. Its also possible that webkit.js is hiding errors since headless-gl doesn't handle 14 out of the 144 functions that webkit.js needs.

Theres also a few forks that have been updated far more recently such as [sbrown345's](https://github.com/sbrown345/webkit.js). Any thoughts more than welcome

**Comment by piranna** [comment]
@formula1, if I understood correctly, have you managed to make `webkit.js` to work against `node-canvas` and Linux framebuffer? Wow... O_O Yeah, it will be slow, but it would be a cool demo and proof of concept, and would be useful until we get to work something more performance savvy, good work! :-D

Could you be able to explain a little bit more how the elements are integrated? :-)

**Comment by Coretool** [comment]
BTW... I managed to make a static build of qt :)... (wasn't that hard) I'll write the rest of it this weekend

**Comment by piranna** [comment]
> BTW... I managed to make a static build of qt :)... (wasn't that hard)

Cool :-) Against what graphic layer, framebuffer or EGL? Have you check it works on your Ubuntu machine launching it from the terminal?

**Comment by Coretool** [comment]
At the moment it is running against the framebuffer (but it is indeed possible to run it against EGL)
I tested ut on a Ubuntu 15 Mate machine (and my raspi). It runs of the terminal...

**Comment by piranna** [comment]
Ah ok, that's a good signal :-) I was worried it would be compiled against X11 or being using some X11 libs. Ok, when you get it tell us so we can review it ;-)

**Comment by lite20** [comment]
> At the moment it is running against the framebuffer (but it is indeed possible to run it against EGL)
I tested ut on a Ubuntu 15 Mate machine (and my raspi). It runs of the terminal...

I want dis tho :3

**Comment by Coretool** [comment]
Hope that I am able to finish it this weekend ^^

**Comment by lite20** [comment]
> Hope that I am able to finish it this weekend ^^

I will camp out and wait!

**Comment by formula1** [comment]
Considering @Coretool is going down a much cleaner approach, i'll probably stop development. I have no idea how webgl works and working within a 15000 line emscripten file isnt too easy.


It turns out that webkit js doesnt use 2d context, only webgl. I havent gotten it to work with frame buffer because I cant render an image : /. I only draw white screens thus far. It doesnt seem to perform too bad (though it doesnt even work in the first place) Id say maybe a solid 20 Frames per second lol

So heres what I did
- webkitjs requires a bunch of globals from the browser - https://github.com/formula1/node-webkit-canvas/blob/master/DOMPolyfills/BrowserSandbox.js
  - window object (which also acts as global)
  - document object (this seems only to be used when going to and from fullscreen, which is still experimental)
  - image - in 2d context, they clone what is written by the another canvas
  - blob - also needed for the other canvas
  - crapload of webgl stuff
  - url to support blob urls likely
- webkit js also needs a canvas as an argument
  - with this canvas it only grabs webgl context

I created or required a polyfill for everything above. Most important being the canvas (which i return headless gl as the webgl context). From there webkitjs does its thing with the globals/settings and thats it.

**Comment by Coretool** [comment]
#### Progress update 
I managed to build a static Qt lib. It is basically the minimal qt lib (without any X-lib-* attachments)... It seems like the examples work fine, so it should be ok. I haven't built the Rendering application yes, because I still have some questions that I would like to have answered before I start writing the 'browser':

Do we need a address bar ? 

Do we need support for svg and gifs ? 

Do we need buttons like 'backward' and 'forward' ? (Or will the desktop implement them ?=

As always, ideas are welcome =)

**Comment by Compynerd255** [comment]
> Do we need an address bar?
> Do we need buttons like 'backward' and 'forward'?

As this is an HTML renderer for the NodeOS desktop, this is just me, but I would say no to these - it seems to me that page history is pretty meaningless here.

But very yes to all the other ones. :)

**Comment by piranna** [comment]
> I managed to build a static Qt lib. It is basically the minimal qt lib
(without any X-lib-* attachments)...

Awesome! :-D

> Do we need a address bar ?

No, we don't need it, since the desktop app will be given as parameter and
we'll not move from there.

> Do we need support for svg and gifs ?

It would be nice to have them specially SVG, but it's ok to enable them in
a future compilation if it give you too much problems.

> Do we need buttons like 'backward' and 'forward' ? (Or will the desktop
implement them ?=

No, they are not needed because it will render the desktop web page given
as parameter, so the concept of navigation history doesn't makes sense here.

**Comment by Coretool** [comment]
Ok, I just finished the first qt build, without svg and gifs. (It takes about an hour to build...)Now I'll move on to build the final application =)

**Comment by piranna** [comment]
> Ok, I just finished the first qt build, without svg and gifs. (It takes about an hour to build...)Now I'll move on to build the final application =)

Ok :+1: please take note of all the steps you are doing and the commands you are using so we can write a script to automate the process and build newer versions later ;-)

**Comment by Coretool** [comment]
Sure ^^ But first I have to find out the right combination of them ;)

**Comment by piranna** [comment]
Well, you can keep writting them anyway so this way we are sure don't miss any magic compiler or configuration flag... ;-)

**Comment by lite20** [comment]
> Do we need a address bar ?

I would assume not seeing that the display manager ought to handle what's shown

> Do we need support for svg and gifs ?

That would be awesome!

> Do we need buttons like 'backward' and 'forward' ? (Or will the desktop implement them ?=

The desktop would need to implement them. If the desktop doesn't want them and they're there its more of a hassle than if they're not there and the desktop wants them.

> Do we need OpenGL ? (maybe for later...)

Once again, would be super awesome to have. Do you mean WebGL or OpenGL?

> Do we need the mouse & keyboard input framebuffer libs (Yes they exist, and I would say yes ^^)

Yes, definitely. Being a graphic layer, a desktop would need these no matter what. Might as well include them!

> Do we need media backends for things like audio ?

This is honestly optional. I don't think we would, as each app would probably have it's own code for managing audio and such. For systems trying to implement multiple rendering platforms (like a yunitto-based Desktop that needs to handle both local, and external connections) it might be pleasant to it to use it's own system.

However if it is included, apps that don't want it could easily ignore it and use their own systems so it's really up to you. :stuck_out_tongue:

**Comment by piranna** [comment]
>> Do we need media backends for things like audio ?
>
> This is honestly optional. I don't think we would, as each app would
probably have it's own code for managing audio and such. For systems trying
to implement multiple rendering platforms (like a yunitto-based Desktop
that needs to handle both local, and external connections) it might be
pleasant to it to use it's own system.
>
> However if it is included, apps that don't want it could easily ignore it
and use their own systems so it's really up to you.

Since graphic apps will be based on html they will use the Audio API, so
yes, se need audio support. Take in account that desktop needs to work on
regular web browsers, too...

**Comment by Coretool** [comment]
# Progress Update

Hey =) I finished the renderer -> it takes the target URL as the line argument and renders to the frame buffer or EGLFS
Where should I dump the code ? @lite20 Do you still want it at Yunitto or should I create a new repo ?

**Comment by piranna** [comment]
> Hey =) I finished the renderer -> it takes the target URL as the line
argument and renders to the frame buffer or EGLFS
> Where should I dump the code ?

What about its own repo :-)

**Comment by Coretool** [comment]
Ok =) I'll create it tomorrow since I don't have a computer near me...

**Comment by Coretool** [comment]
The code is now online, go visit it [here](http://github.com/coretool/noGui). I named the project *noGui* which stands for **n**ode**O**s-**Gui**. Waiting for your comments on it \^_^/

**Comment by piranna** [comment]
> I named the project noGui which stands for nodeOs-Gui.

I love it :-D I wanted to use "nodeos-" as prefix for NodeOS related
modules and projects, but I like the idea of using "no" as short
alternative or for projects not so directly related but still focused for
Node.js, similar to how "tx" is used as prefix for Python Twisted packages
:-D

**Comment by Coretool** [comment]
Cool that you like it ^^

**Comment by piranna** [comment]
> Cool that you like it ^^
>
It was a nice idea :-)

**Comment by lite20** [comment]
@coretool you can create a new repo or put it in Yunitto. Your choice. I'll just cross-clone it into Yunitto if you put it into another repo ^_^

**Comment by piranna** [comment]
> @coretool you can create a new repo or put it in Yunitto. Your choice.
I'll just cross-clone it into Yunitto if you put it into another repo ^_^

It's on the noGUI repo. Why do you want to copy it on Yunitto? Is not
possible to have it as dependency, or download it automatically?

**Comment by Coretool** [comment]
Everything is possible :)

**Comment by Coretool** [comment]
@lite20 No, it hasn't been modulized yet, but I am working on that (automated build script(s) are in development, thanks to piranna).

**Comment by piranna** [comment]
> @lite20 No, it hasn't been modulized yet, but I am working on that
(automated build script(s) are in development, thanks to piranna).

You are welcome :-)

**Comment by lite20** [comment]
> It's on the noGUI repo. Why do you want to copy it on Yunitto? Is not
possible to have it as dependency, or download it automatically?

I was under the impression that it wasn't yet a module. But since you have build script(s) coming, integration will be easy.

**Comment by lite20** [comment]
Just if anyone was wondering, Yunitto's browser rendering works and is open for experimentation. If anyone has issues with it, please let me know so we can clean it up

**Comment by piranna** [comment]
Could you be able to remind us what was the purposse of Yunitto, please?

**Comment by lite20** [comment]
@piranna it's designed for desktop managers. (Works on any platform including NodeOS) and handles multiple screens and interfaces for the DM. for example it provides an easy way for a DM to support rendering to a local screen (using noGUI), or a browser over the same interface.

**Comment by lite20** [comment]
Can we pin this thread?

**Comment by piranna** [comment]
Done. Don't you have permissions to add labels?

**Comment by lite20** [comment]
@piranna yes, I do :P I just wanted to ask first

**Comment by piranna** [comment]
Cool ;-) Don't need to ask for this small things :-)
El 1/7/2016 8:19, "Lite McFish" <notifications@github.com> escribió:

@piranna <https://github.com/piranna> yes, I do :P I just wanted to ask
first

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/NodeOS/NodeOS/issues/201#issuecomment-229863240>, or mute
the thread
<https://github.com/notifications/unsubscribe/AAgfvhIJI0gSOyekctCCAzpianVY_N5Pks5qRLGBgaJpZM4G-r60>
.

**Comment by lite20** [comment]
Hahah okay, I won't ask next time :P

**Comment by piranna** [comment]
> Hahah okay, I won't ask next time :P

;-)

**Comment by cazzer** [comment]
I've been following this project for a few years and am excited about the GUI prospects, is this still the right thread for such a discussion, (#311 seems more active)? Either way, I have a couple of comments and questions:

@Coretool Just wanted to mention, `nogui` is not very searchable on the Google machine, but I did find your package by searching "github nodeos-gui". [Here it is, for anyone else looking.](https://github.com/Coretool/noGUI).

@piranna I've seen in the mention of QT in other threads, (like issue #311), as well as WebKit and Blink. I'm not familiar with the C++ layer required to make rendering a reality, but where are we in relation to the checklist at the top of this thread?

I've seen recurring mention to #147 so I also want to express my viewpoint: This project should implement, or provide a platform for, Blink, or a similar *external* dependency on which developers can implement their own visual layer, but not impose any opinions on the view layer itself. As an example: "GUI windows" are mention in #147, but that raises concern that this platform will result in the same sort of application layer we get from conventional operation systems, (which some users may not see as an improvement).

**TL;DR How do we get to a web browser-like environment from where we're at now?** The flexibility of a blank page running this much closer to the hardware sounds like a damn good place to build an interface, regardless of the hardware powering it.

p.s. @piranna and contributors: This project is amazing, thank you for your commitment!

**Comment by formula1** [comment]
@cazzar I believe that the graphics has been one of the hardest points for many of us. While in concept it is simple, in practice it has turned up not so much. However, passionate people like me, you, @Coretool, @lite20 and of course @piranna keep the vision alive.

I believe the next steps to getting this done is isolating what we need for graphics.

1. A way to render to Hardware - Using framebuffer we only need to provide it RGB values
2. A way to transform HTML to RGB Values
  - Talk to someone from the blink/webkit team to see how we can incorporate their rendering engine
    - Chromium Embedded Framework was a thing for a period of time
    - Opera started using blink as well so its not unheard of or unreasonable
  - Talk to someone on servo - Rust + Javascript may be a future we can get into
  - Ideally we use blink or some other rendering engine, however I know when I look at the code I certainly get fatigued just
  - Edit [phantomJS](http://phantomjs.org/screen-capture.html) to see if we can do streaming output or capture on each dom change (avoid css transformations)
  - Edit a pre-existing html to renderable output such as https://www.npmjs.com/package/wkhtmltopdf
  - machine learn our way to a rendering engine :)
3. A way to get the code into the rendering system and out again
  - The easy part as we can just create a wrapper library

I think we are all busy people so the more hands that can dig into this the better. I myself have about 3 projects I currently am taking care of though I still keep an eye on this one because its special to me

**Comment by lite20** [comment]
@formula1 I have an idea which could provide a rough but experimental graphic runtime for NodeOS. :P I'll keep you all posted

**Comment by Technohacker** [comment]
@piranna Mesa can be compiled without X dependencies by disabling GLX and XA. However, the main problem is that some components of Mesa (GBM and EGL) cannot be built as static libraries.

Another issue I've faced is that on compiling Mesa with the NodeOS toolchain, it failed on a missing zlib.h

**Comment by piranna** [comment]
> @piranna Mesa can be compiled without X dependencies by disabling GLX and XA. However, the main problem is that some components of Mesa (GBM and EGL) cannot be built as static libraries.

This components are mostly related to hardware drivers, so it makes sense that can't be compiled statically, but instead each system should have their own ones for their particular hardware. We already have `libfuse` as a dynamic library because it's fairly common and had some problems compiling it statically, so I don't see it bad to add this ones too, only that stick to the really few needed ones, and if possible put them in the user directory instead of system globally.


> Another issue I've faced is that on compiling Mesa with the NodeOS toolchain, it failed on a missing zlib.h

Possibly related to https://github.com/Coretool/noGUI/issues/10.
